src/v/kafka/protocol/tests/protocol_test.cc

```diff
@@ -186,8 +186,145 @@ void check_all_requests(kafka::type_list<Ts...>) {
     (check_proto_compat<Ts>(), ...);
 }
 
+template<typename Request, api_version::type version, bool is_request>
+requires(KafkaApiHandler<Request>)
+struct tag_field_entry {
+    using api = Request::api;
+    static constexpr api_version test_version = api_version(version);
+    static constexpr bool request = is_request;
+};
+
+// Instructions on adding optional tag field messages
+// When a new kafka message is added that contains optional tag fields,
+// you will do the following:
+// 1. Add the entry to `tag_field_entries`, providing
+//    a. The handler
+//    b. minimum api version
+//    c. true for request, false for response (needs to be bool for constexpr)
+// 2. Create a specialized `create_default_and_non_default_data` function
+//    a. Have it set non-default values and default values for each
+//    b. Provide the difference in size the encoded buffers will be
+
+using tag_field_entries = kafka::type_list<
+  tag_field_entry<create_topics_handler, 5, false>,
+  tag_field_entry<api_versions_handler, 3, false>>;
+
+template<typename T>
+long create_default_and_non_default_data(T& non_default_data, T& default_data);
+
+template<>
+long create_default_and_non_default_data(
+  decltype(create_topics_response::data)& non_default_data,
+  decltype(create_topics_response::data)& default_data) {
+    non_default_data.throttle_time_ms = std::chrono::milliseconds(1000);
+
+    non_default_data.topics.emplace_back(creatable_topic_result{
+      model::topic{"topic1"},
+      {},
+      kafka::error_code{1},
+      "test_error_message",
+      3,
+      16,
+      std::nullopt,
+      kafka::error_code{2}});
+
+    default_data = non_default_data;
+
+    default_data.topics.at(0).topic_config_error_code = kafka::error_code{0};
+
+    // int16 (2 bytes) + tag (2 bytes)
+    return 4;
+}
+
+template<>
+long create_default_and_non_default_data(
+  decltype(api_versions_response::data)& non_default_data,
+  decltype(api_versions_response::data)& default_data) {
+    non_default_data.finalized_features_epoch = 0;
+    default_data = non_default_data;
+    default_data.finalized_features_epoch = -1;
+
+    // int64 (8 bytes) + tag (2 bytes)
+    return 10;
```

Comment:
maybe its better to just serialize to iobuf and return `size_bytes`, in case theres future regressions that would catch an issue

Reply:
Not sure I'm following here, let's sync on this